### PR TITLE
compilers: Embed symbols in object files when using MSVC

### PR DESCRIPTION
**Pull request by oleavr**
Otherwise each of any static library's object files will be referring to
its .pdb, making redistribution very impractical.

Executables and DLLs still get a .pdb each, though.

Other possible solutions:

- Merge each object-file's `.pdb` into one big one next to the `.a`.
  I'm not aware of any tool able to do this, so this might be a lot of
  work to implement.
- Feed the compiler with all of the C-files for the target being built,
  so it can output one `.pdb` for all of them. This breaks down when
  using `.extract_objects()`.

So bottom line is that I'm not happy with this solution, but I'm opening
this PR to start the brainstorming. Hopefully there's a better option I
haven't thought of.

### CODE REVIEW

**Comment by jpakkane** [comment]
I seem to remember (possibly incorrectly) that the use of `/Z7` is either deprecated or strongly discouraged by Microsoft. Do you know if this is actually the case? Is that flag guaranteed to keep working?

Somewhat relatedly, we might, at some point, have to add support for split dwarf info as well. I have not looked at the details on how that is implemented, but we might need to do a global toggle "have split debug info" or something like that that would then map to the corresponding platform flags. Dunno if the implementation on MSVC and GCC is close enough to map these to the same thing, however.

**Comment by nirbheek** [comment]
AFAIK: `/Z7` (embedded debug info) doesn't provide advanced debug features like edit and continue and contains less information than PDBs. `/Z7` also leads to larger executables in general, and is a pain to use with symbol servers. We should definitely not switch the default from `/ZI` to `/Z7`.

Everyone seems to use PDBs these days, but I have been wondering whether we should just provide an option to switch between pdb and embedded debug info. This can then be combined with `override_options:` to set it per-target and fix the problem described in this PR.

**Comment by oleavr** [comment]
> the use of /Z7 is either deprecated or strongly discouraged by Microsoft

That's also my understanding and why I mentioned I'm not happy with this solution.

> /Z7 also leads to larger executables in general

FWIW that's not an issue with this PR, as symbols are only embedded in object-files and static libraries, whereas PDBs are still produced for EXEs and DLLs.

That being said, I really don't like this solution at all, both because it's using a legacy feature and because it's conflicting with edit and continue etc. I'm just not sure how to solve this properly, so this is just the band-aid I'm using for Frida's needs until we come up with something better. (The use-case there is producing a prebuilt SDK consisting of just static libraries and header files, where the consumer of that SDK wants to use it without missing out on debug symbols.)

**Comment by nirbheek** [comment]
> > /Z7 also leads to larger executables in general
> 
> FWIW that's not an issue with this PR, as symbols are only embedded in object-files and static libraries,

The larger executable size is because the debug information size is larger in the object files, so it does affect. Although that's just a trade-off and not a blocker.

**Comment by jpakkane** [comment]
Maybe there should be an option between embedded and separate debug info, because you can do split debug also on Linux with `split-dwarf` or some compiler flag like that. IIRC people building Qt or other large programs seem to want it.

**Comment by jpakkane** [comment]
[Relevant link.](http://www.productive-cpp.com/improving-cpp-builds-with-split-dwarf/)

**Comment by oleavr** [comment]
@jpakkane That would be awesome!
